### ISSUE TITLE
recursive function crashes

### ISSUE THREAD

**Issue by bakul** [kevinlawler/kona#393]
Permutation function (cribbed from nsl.com) crashes on second invocation for any x > 1
````
  p:{:[1<x;,/(>:'(x,x)#1,x#0)[;0,'1+_f x-1];,!x]}
  p 3
  p 3
````

**Comment by kirbyfan64** [comment]
gdb says:

```
ryan@DevPC-LX:~$ rlwrap gdb k
GNU gdb (Ubuntu 7.7.1-0ubuntu5~14.04.2) 7.7.1
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from k...done.
Reading symbols from k...
Starting program: /media/ryan/stuff/bin/k 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
K Console - Enter \ for help

  [New Thread 0x7f233971d700 (LWP 13105)]

    p:{:[1<x;,/(>:'(x,x)#1,x#0)[;0,'1+_f x-1];,!x]}
{:[1<x;,/(>:'(x,x)#1,x#0)[;0,'1+_f x-1];,!x]}
  p 3
(0 1 2
 0 2 1
 1 0 2
 1 2 0
 2 0 1
 2 1 0)
  p 3

Program received signal SIGSEGV, Segmentation fault.
0x0000000000417bf1 in ex2 (v=<optimized out>, k=<optimized out>)
    at src/kx.c:858
858	    if(prnt && z->t==7) {
(gdb) bt
#0  0x0000000000417bf1 in ex2 (v=<optimized out>, k=<optimized out>)
    at src/kx.c:858
#1  0x0000000000417de1 in ex2 (v=0x7f233a427a98, k=0x0) at src/kx.c:913
#2  0x000000000041adc2 in ex0 (v=0x7f233a427a98, k=<optimized out>, 
    r=<optimized out>) at src/kx.c:629
#3  0x000000000041902b in ex_ (a=<optimized out>, r=4) at src/kx.c:598
#4  0x00000000004179e4 in ex2 (v=0x7f233a424bd8, k=<optimized out>)
    at src/kx.c:839
#5  0x000000000041af3f in ex0 (v=0x7f233a424bd8, k=<optimized out>, 
    r=<optimized out>) at src/kx.c:620
#6  0x000000000041902b in ex_ (a=<optimized out>, r=0) at src/kx.c:598
#7  0x000000000041704c in ex (
    a=<error reading variable: Cannot access memory at address 0x0>)
    at src/kx.c:606
#8  vf_ex (q=<optimized out>, g=<optimized out>) at src/kx.c:505
#9  0x000000000041a76e in dv_ex (a=<optimized out>, p=<optimized out>, 
    b=<optimized out>) at src/kx.c:340
#10 0x0000000000418358 in ex2 (v=0x7f233a4278c8, k=<optimized out>)
    at src/kx.c:960
#11 0x0000000000417de1 in ex2 (v=0x7f233a4278b8, k=0x0) at src/kx.c:913
#12 0x0000000000417de1 in ex2 (v=0x7f233a4278a0, k=0x0) at src/kx.c:913
#13 0x000000000041b0e0 in ex0 (v=0x7f233a427898, k=0x0, r=2) at src/kx.c:652
#14 0x000000000041902b in ex_ (a=<optimized out>, r=2) at src/kx.c:598
---Type <return> to continue, or q <return> to quit---
#15 0x0000000000418fcf in ex_ (a=<optimized out>, r=1) at src/kx.c:593
#16 0x0000000000417bbe in ex2 (v=<optimized out>, k=<optimized out>)
    at src/kx.c:857
#17 0x0000000000417d20 in ex2 (v=0x7f233a4274b8, k=0x0) at src/kx.c:937
#18 0x000000000041b50a in ex0 (v=0x7f233a427498, k=<optimized out>, 
    r=<optimized out>) at src/kx.c:629
#19 0x000000000041902b in ex_ (a=<optimized out>, r=4) at src/kx.c:598
#20 0x00000000004179e4 in ex2 (v=0x7f233a42ac98, k=<optimized out>)
    at src/kx.c:839
#21 0x000000000041af3f in ex0 (v=0x7f233a42ac98, k=<optimized out>, 
    r=<optimized out>) at src/kx.c:620
#22 0x000000000041902b in ex_ (a=<optimized out>, r=0) at src/kx.c:598
#23 0x000000000041704c in ex (
    a=<error reading variable: Cannot access memory at address 0x0>)
    at src/kx.c:606
#24 vf_ex (q=<optimized out>, g=<optimized out>) at src/kx.c:505
#25 0x000000000041a76e in dv_ex (a=<optimized out>, p=<optimized out>, 
    b=<optimized out>) at src/kx.c:340
#26 0x0000000000418358 in ex2 (v=0x7f233a421998, k=<optimized out>)
    at src/kx.c:960
#27 0x000000000041af3f in ex0 (v=0x7f233a421998, k=<optimized out>, 
    r=<optimized out>) at src/kx.c:620
#28 0x000000000041902b in ex_ (a=<optimized out>, r=0) at src/kx.c:598
---Type <return> to continue, or q <return> to quit---
#29 0x0000000000418e7e in ex (
    a=<error reading variable: Cannot access memory at address 0x0>)
    at src/kx.c:606
#30 0x00000000004153f1 in line (f=<optimized out>, a=<optimized out>, 
    n=0x7ffd204c5848, p=<optimized out>) at src/kc.c:244
#31 0x0000000000415f3e in attend () at src/kc.c:388
#32 0x000000000042a807 in main (argc=0, argv=0x1) at src/main.c:9
   quit
A debugging session is active.

	Inferior 1 [process 13101] will be killed.

Quit anyway? (y or n) y
ryan@DevPC-LX:~$ 
```

**Comment by tavmem** [comment]
This is a regression error, and has a history of problems.
In the commit 916c097275a37e1fd5bbd5b5ecbe00d9db9666a0 dated Dec 29, 2013, titled "cosmetics" it works:
```
  p:{:[1<x;,/(>:'(x,x)#1,x#0)[;0,'1+_f x-1];,!x]};
  p 2
(0 1
 1 0)
  p 2
(0 1
 1 0)
```
In the commit d66897c23a79a6f9c916d6122b04e20584a11b6b dated Dec 30, 2013, titled "fix #221 -- (3rd attempt)", it crashes:
```
  p:{:[1<x;,/(>:'(x,x)#1,x#0)[;0,'1+_f x-1];,!x]};
  p 2
(0 1
 1 0)
  p 2
rlwrap: warning: k killed by SIGSEGV.
```
In the commit c75f71e4f42ce6cf63ec0e85b601564525058687 dated Dec 30, 2013, titled "fix #227 Silencing errors causes crash" it yields "undescribed error":
```
  p:{:[1<x;,/(>:'(x,x)#1,x#0)[;0,'1+_f x-1];,!x]};
  p 2
(0 1
 1 0)
  p 2
undescribed error
```
In commit 0014127b3a8008c742301ba16ea64a7bea78e363 dated Feb 26, 2014, titled "partial fix of issue 238", it works again:
```
  p:{:[1<x;,/(>:'(x,x)#1,x#0)[;0,'1+_f x-1];,!x]};
  p 2
(0 1
 1 0)
  p 2
(0 1
 1 0)
```
In commit 246b18c570ed932fb4224f2d7c680e3bc9a8b07d dated Jun 19, 2015, titled "issue 127: cut timing from 33 sec to 14 sec" it crashes again:
```
  p:{:[1<x;,/(>:'(x,x)#1,x#0)[;0,'1+_f x-1];,!x]};
  p 2
(0 1
 1 0)
  p 2
rlwrap: warning: k killed by SIGSEGV.
```
